firedrake/checkpointing.py

```diff
@@ -716,6 +716,41 @@ def _save_mesh_topology(self, tmesh):
             perm_is.setName(None)
             self.viewer.popGroup()
 
+    @PETSc.Log.EventDecorator("SetTimstep")
+    def _set_timestep(self, mesh, name, idx, time):
+        # follow the general path rule for storing a function
+        tmesh_name = self._get_mesh_name_topology_name_map()[mesh.name]
+        V_name = self._get_function_name_function_space_name_map(tmesh_name, mesh.name)[name]
+        path = self._path_to_function(tmesh_name, mesh.name, V_name, name)
+        # check if the function's path has the attribute
+        if self.has_attr("/"+path if path else "", "stored_time_steps"):
+            # collect previous indices and timesteps and add the current timestep path to it
+            old_indices, old_times = self.get_timesteps(mesh, name)
+            times = np.concatenate((old_times, [time]))
+            indices = np.concatenate((old_indices, [len(old_indices) if idx is None else idx]))
+        else:
+            # initiate timestepping, assuming the saving the function has succeeded
+            indices = [0 if idx is None else idx]
+            times = [time]
+        # store all timesteps and indices
+        self.set_attr("/"+path if path else "", "stored_time_indices", indices)
+        self.set_attr("/"+path if path else "", "stored_time_steps", times)
+
+    @PETSc.Log.EventDecorator("GetTimsteps")
+    def get_timesteps(self, mesh, name):
```

Comment:
Please provide docstrings. Ideally in [numpydoc format](https://numpydoc.readthedocs.io/en/latest/format.html).

Reply:
Thanks. Added.